// adaptivejs/tasks/config/sync.js
module.exports = function(grunt) {
    var generateComponentConfigs = function() {
        var componentLocationsJson = 'app/config/component.json';

        var componentLocations = grunt.file.readJSON(componentLocationsJson);

        var localPath = 'app/' + componentLocations['local'] + '/*/assets';
        var installedPath = 'app/' + componentLocations['stencil'] + '/stencil-*/assets';

        var localComponents = grunt.file.expand({ filter: 'isDirectory' }, localPath);
        var installedStencils = grunt.file.expand({ filter: 'isDirectory' }, installedPath);

        return localComponents.concat(installedStencils).map(function(path) {
            return {
                cwd: path,
                src: ['**'],
                dest: 'build/assets'
            };
        });
    };

    var syncFileConfigs = generateComponentConfigs();
    var staticConfig = { cwd: 'static', src: ['**'], dest: 'build/' };

    syncFileConfigs.push(staticConfig);

    return {
        assets: {
            files: syncFileConfigs,
            verbose: false
        }
    };
};
