// adaptivejs/tasks/config/sync.js
module.exports = function(grunt) {
    var generateComponentConfigs = function() {
        var componentLocationsJson = 'app/config/component.json';
        var componentLocations = grunt.file.readJSON(componentLocationsJson);

        var localPath = 'app/' + componentLocations['local'] + '/*/assets';
        var installedPath = 'app/' + componentLocations['stencil'] + '/stencil-*/assets';

        var localComponents = grunt.file.expand({ filter: 'isDirectory' }, localPath);
        var installedStencils = grunt.file.expand({ filter: 'isDirectory' }, installedPath);

        return localComponents.concat(installedStencils).map(function(path) {
            return {
                cwd: path,
                src: ['**'],
                dest: 'build/assets'
            };
        });
    };

    var staticConfig = [{ cwd: 'static', src: ['**'], dest: 'build/' }];

    if (grunt.file.isFile('app/config/component.json')) {
        var syncFileConfigs = generateComponentConfigs();
        staticConfig.push(syncFileConfigs);
    }

    return {
        assets: {
            files: staticConfig,
            verbose: false
        }
    };
};
